### ISSUE TITLE
FreeBSD build broken due to -ldl flag

### ISSUE THREAD

**Issue by clemensg** [phusion/passenger#1805]
**Question 1: What is the problem?**

Building passenger 5.0.28/nginx fails on FreeBSD 10.3-RELEASE with the following error message:

```
...
c++ -o buildout/support-binaries/PassengerAgent buildout/common/libpassenger_common/Logging.o buildout/common/libpassenger_common/Exceptions.o buildout/common/libpassenger_common/Utils/SystemTime.o buildout/common/libpassenger_common/Utils/StrIntUtils.o buildout/common/libpassenger_common/Utils/StrIntUtilsNoStrictAliasing.o buildout/common/libpassenger_common/Utils/IOUtils.o buildout/common/libpassenger_common/Utils.o buildout/common/libpassenger_common/vendor-modified/modp_b64.o buildout/common/libpassenger_common/jsoncpp.o buildout/common/libpassenger_common/UnionStationFilterSupport.o buildout/common/libpassenger_common/Utils/CachedFileStat.o buildout/common/libpassenger_common/Utils/LargeFiles.o buildout/common/libpassenger_common/MemoryKit/mbuf.o buildout/common/libpassenger_common/MemoryKit/palloc.o buildout/common/libpassenger_common/ServerKit/http_parser.o buildout/common/libpassenger_common/ServerKit/Implementation.o buildout/common/libpassenger_common/DataStructures/LString.o buildout/common/libpassenger_common/Utils/Hasher.o buildout/common/libpassenger_common/AppTypes.o buildout/support-binaries/AgentMain.o buildout/support-binaries/AgentBase.o buildout/support-binaries/WatchdogMain.o buildout/support-binaries/CoreMain.o buildout/support-binaries/CoreApplicationPool.o buildout/support-binaries/CoreController.o buildout/support-binaries/UstRouterMain.o buildout/support-binaries/SystemMetricsMain.o buildout/support-binaries/TempDirToucherMain.o buildout/support-binaries/SpawnPreparerMain.o buildout/common/libboost_oxt.a   buildout/libev/.libs/libev.a  -lm buildout/libuv/.libs/libuv.a  -lrt -lpthread -lkvm -L/usr/local/lib -lcurl -lz -lpthread -lrt -ldl
/usr/bin/ld: cannot find -ldl
c++: error: linker command failed with exit code 1 (use -v to see invocation)
rake aborted!
Command failed with status (1): [c++ -o buildout/support-binaries/PassengerAgent buildout/common/libpassenger_common/Logging.o buildout/common/libpassenger_common/Exceptions.o buildout/common/libpassenger_common/Utils/SystemTime.o buildout/common/libpassenger_common/Utils/StrIntUtils.o buildout/common/libpassenger_common/Utils/StrIntUtilsNoStrictAliasing.o buildout/common/libpassenger_common/Utils/IOUtils.o buildout/common/libpassenger_common/Utils.o buildout/common/libpassenger_common/vendor-modified/modp_b64.o buildout/common/libpassenger_common/jsoncpp.o buildout/common/libpassenger_common/UnionStationFilterSupport.o buildout/common/libpassenger_common/Utils/CachedFileStat.o buildout/common/libpassenger_common/Utils/LargeFiles.o buildout/common/libpassenger_common/MemoryKit/mbuf.o buildout/common/libpassenger_common/MemoryKit/palloc.o buildout/common/libpassenger_common/ServerKit/http_parser.o buildout/common/libpassenger_common/ServerKit/Implementation.o buildout/common/libpassenger_common/DataStructures/LString.o buildout/common/libpassenger_common/Utils/Hasher.o buildout/common/libpassenger_common/AppTypes.o buildout/support-binaries/AgentMain.o buildout/support-binaries/AgentBase.o buildout/support-binaries/WatchdogMain.o buildout/support-binaries/CoreMain.o buildout/support-binaries/CoreApplicationPool.o buildout/support-binaries/CoreController.o buildout/support-binaries/UstRouterMain.o buildout/support-binaries/SystemMetricsMain.o buildout/support-binaries/TempDirToucherMain.o buildout/support-binaries/SpawnPreparerMain.o buildout/common/libboost_oxt.a   buildout/libev/.libs/libev.a  -lm buildout/libuv/.libs/libuv.a  -lrt -lpthread -lkvm -L/usr/local/lib -lcurl -lz -lpthread -lrt -ldl  ]
/usr/local/rvm/gems/ruby-2.3.1/gems/passenger-5.0.28/build/support/cplusplus.rb:52:in `run_compiler'
/usr/local/rvm/gems/ruby-2.3.1/gems/passenger-5.0.28/build/support/cplusplus.rb:116:in `create_cxx_executable'
/usr/local/rvm/gems/ruby-2.3.1/gems/passenger-5.0.28/build/agent.rb:81:in `block in <top (required)>'
/usr/local/rvm/gems/ruby-2.3.1/bin/ruby_executable_hooks:15:in `eval'
/usr/local/rvm/gems/ruby-2.3.1/bin/ruby_executable_hooks:15:in `<main>'
Tasks: TOP => nginx => nginx_without_native_support => buildout/support-binaries/PassengerAgent
...
```

Looks like it is trying to link to libdl which is not needed and does not exist on FreeBSD.

**Question 2: Passenger version and integration mode**:
Open source 5.0.28/nginx 1.10.0

**Question 3: OS or Linux distro, platform (including version)**:
FreeBSD 10.3-RELEASE, x86_64
FreeBSD clang version 3.4.1

**Question 4: Passenger installation method**:
[x] RubyGems, no Gemfile

**Question 5: Your app's programming language (including any version managers) and framework (including versions)**:
Ruby 2.3.1, RVM, Rails

**Question 6: Are you using a PaaS and/or containerization? If so which one?**
No!

**Question 7: Anything else about your setup that we should know?**
No!

**Comment by FooBarWidget** [comment]
@OnixGH I think the LVE patch broke this. The build system should only add `-ldl` if the system actually has dl. There is a PlatformInfo check for this. If you want to test, the easiest way to spin up a FreeBSD box is probably through Digital Ocean.

**Comment by clemensg** [comment]
@FooBarWidget If you have an idea what commit should be reverted, I can test it for you.

I tried fixing it myself but I could only find `-ldl` in the gyp files for libuv, however it is only added conditionally for the Linux and Android build, right?

Can you tell where this flag comes from?

**Comment by OnixGH** [comment]
Reproduced on FreeBSD 1.2 (D.O). I'll have a look at the LVE patch in the light of this later on.

**Comment by FooBarWidget** [comment]
In src/ruby_supportlib/phusion_passenger/platform_info/cxx_portability.rb, method portability_c_or_cxx_ldflags:

    result << ' -ldl'

There should be a check there that only adds this flag if there actually is a 'dl' library.

**Comment by clemensg** [comment]
Ah, I searched through that file myself before, but forgot to switch to the stable-5.0 branch, was still at master.... I wondered why there is no -ldl flag ;-)
Anyway, thank you!

Fixed by commit dbc3e41852efb763b5d6a18840204407af0cf40f

**Comment by CirrusThink** [comment]
We are having same issue when deploying 5.0.28 enterprise edition, how do we patch this ? FreeBSD 10.3

**Comment by CirrusThink** [comment]
manually patched for now worked

**Comment by FooBarWidget** [comment]
Manually patching is fine for now. We will incorporate this fix in the next release, 5.0.29.
